extensions/amp-story/1.0/amp-story.js

```diff
@@ -1285,15 +1302,30 @@ export class AmpStory extends AMP.BaseElement {
 
     if (
       isExperimentOn(this.win, 'amp-story-paywall-exp') &&
-      targetPage.isPaywallProtected()
+      targetPage.isPaywallProtected() &&
+      !this.granted_
```

Comment:
I should not be able to navigate to a potentially protected page till the entitlements are fetched and the grant status up to date

Reply:
Do you mean we should fetch the grant status again here to avoid the case where the grantStatus changes between the time user loads the story and the time user hits the paywall story page? I think this should happen very rarely only when the user signs up somewhere else and the publisher's system is not up to date yet until the user hits the paywall story page.

But I can add one line here. Just to make sure I understand your comment

Reply:
Going off of this and other comments in the file, what if we just make the subscription state part of the store? I think we could get rid of the `SubscriptionService` and any granted state property in this file. This also may help clean up the the overall logic and simplify testing

```js
 const SubscriptionState = {
  BLOCKED: 1,
  GRANTED: 2,
};

// amp-story.js 

if (!targetPage.isPaywallProtected()) {
  this.storeService_.dispatch(Action.HIDE_UI);
}

const subscriptionState = this.storeService_.get(
  StateProperty.SUBSCRIPTIONS_STATE
);
if (!subscriptionState.GRANTED && targetPage.isPaywallProtected()) {
  this.storeService_.dispatch(Action.SHOW_UI);
  return Promise.resolve();
}

// amp-story-subscription.js

this.storeService_.subscribe(UI_STATE, (state) => {
  // hide/show UI are where css classes are actually updated
  state ? hideUI() : showUI();
});

// All communication with subcriptionService in this file
SubscriptionService.whenGranted(() => {
  this.storeService_.dispatch(
    Action.UPDATE_SUBS_STATE,
    SubscriptionState.GRANTED
  );

  this.storeService_.dispatch(Action.SHOW_UI, false);
});
```

Reply:
Just realized this is basically the same idea y'all are alluding to in the other comments :)

Reply:
Yep exactly : ) And it's done now.